libs/libc/unistd/lib_execle.c

```diff
@@ -0,0 +1,211 @@
+ *   'execv()' suffer from some incompatibilities that may or may not be
+ *   addressed in a future version of NuttX.  Other than just being an
+ *   inefficient use of MCU resource, the most serious of these is that
+ *   the exec'ed task will not have the same task ID as the vfork'ed
+ *   function.  So the parent function cannot know the ID of the exec'ed
+ *   task.
+ *
+ * Input Parameters:
+ *   path - The path to the program to be executed.  If CONFIG_LIBC_ENVPATH
+ *     is defined in the configuration, then this may be a relative path
+ *     from the current working directory.  Otherwise, path must be the
+ *     absolute path to the program.
+ *   ... - A list of the string arguments to be recevied by the
+ *     program.  Zero indicates the end of the list.
+ *
+ * Returned Value:
+ *   This function does not return on success.  On failure, it will return
+ *   -1 (ERROR) and will set the 'errno' value appropriately.
+ *
+ ****************************************************************************/
+
+int execle(FAR const char *path, FAR const char *arg0, ...)
+{
+  FAR char **argv = NULL;
+  FAR char **envp = NULL;
+  FAR char *arg = (FAR char *)arg0;
+  size_t nargs;
+  size_t nenvs;
+  va_list ap;
+  size_t i;
+  int ret;
+
+  /* Count the number of arguments and environment variables */
+
+  va_start(ap, arg0);
+  nargs = 0;
+
+  while (arg != NULL)
+    {
+      /* Yes.. increment the number of arguments.  Here is a sanity
+       * check to prevent running away with an unterminated argv[] list.
+       * MAX_EXECL_ARGS should be sufficiently large that this never
+       * happens in normal usage.
+       */
+
+      if (++nargs > MAX_EXECL_ARGS)
+        {
+          set_errno(E2BIG);
+          va_end(ap);
+          return ERROR;
+        }
+
+      arg = va_arg(ap, FAR char *);
+    }
+
+  arg = va_arg(ap, FAR char *);
+  nenvs = 0;
+
+  while (arg != NULL)
+    {
+      /* Yes.. increment the number of environment variables.  Here is a
+       * sanity check to prevent running away with an unterminated envp[]
+       * list. MAX_EXECL_ENVS should be sufficiently large that this never
+       * happens in normal usage.
+       */
+
+      if (++nenvs > MAX_EXECL_ENVS)
+        {
+          set_errno(E2BIG);
+          va_end(ap);
+          return ERROR;
+        }
+
+      arg = va_arg(ap, FAR char *);
+    }
+
+  va_end(ap);
+
+  /* Allocate a temporary argv[] and envp[] array */
+
+  argv = (FAR char **)lib_malloc((nargs + 1) * sizeof(FAR char *));
+  if (argv == NULL)
+    {
+      set_errno(ENOMEM);
+      return ERROR;
+    }
+
+  envp = (FAR char **)lib_malloc((nenvs + 1) * sizeof(FAR char *));
+  if (envp == NULL)
+    {
+      lib_free(argv);
+      set_errno(ENOMEM);
+      return ERROR;
+    }
+
+  /* Collect the arguments into the argv[] and envp[] array */
+
+  va_start(ap, arg0);
+
+  argv[0] = (FAR char *)arg0;
+  for (i = 1; i <= nargs; i++)
+    {
+      argv[i] = va_arg(ap, FAR char *);
+    }
+
```

Comment:
Should we `argv[nargs] = NULL;`? Same for `envp`

Reply:
Don't need since the code call va_arg until it return NULL.